### PR TITLE
Add new syntheticDevice id column to rewards and implement calculations, add tests

### PR DESCRIPTION
**Pull request by 0xdev22**
# Proposed Changes

### CODE REVIEW

File: internal/services/rewards.go [hunk 1]
```diff
@@ -229,6 +231,28 @@ func (t *BaselineClient) assignPoints() error {
...
+				thisWeek.SyntheticDeviceID = null.IntFrom(int(sd.TokenId))
+				thisWeek.IntegrationPoints += int(integr.Points)
+				thisWeek.IntegrationIds = append(thisWeek.IntegrationIds, integr.Id)
+			}
+
+		} else {
+			logger.Warn().Msgf("No SyntheticDevice Connection for vehicle with ID %s", ud.Id)
+		}
+
 		// Check software integrations.
```
Comment (elffjs) [inline internal/services/rewards.go:256]: We should get rid of this section, yeah? We're replacing it.

Comment (elffjs) [reply internal/services/rewards.go:256]: And if we do delete it, we can probably dump `swIntegrsByID`.

File: internal/services/rewards.go [hunk 2]
```diff
@@ -229,6 +231,28 @@ func (t *BaselineClient) assignPoints() error {
...
+				logger.Warn().Msgf("Synthetic device IntegrationTokenID %d does not have an associated integration.", sd.IntegrationTokenId)
+				continue
+			}
+
+			if integsSignalsThisWeek.Contains(integr.Id) {
+				thisWeek.SyntheticDeviceID = null.IntFrom(int(sd.TokenId))
+				thisWeek.IntegrationPoints += int(integr.Points)
+				thisWeek.IntegrationIds = append(thisWeek.IntegrationIds, integr.Id)
+			}
+
```
Comment (elffjs) [inline internal/services/rewards.go:251]:
```suggestion
```

File: internal/services/rewards.go [hunk 3]
```diff
@@ -229,6 +231,28 @@ func (t *BaselineClient) assignPoints() error {
 			}
 		}
 
+		if sd := ud.SyntheticDevice; sd != nil {
+			if sd.IntegrationTokenId == 0 {
+				logger.Warn().Msgf("Integration with tokenID %d has not been minted.", sd.IntegrationTokenId)
```
Comment (elffjs) [inline internal/services/rewards.go:236]: I would say, "synthetic device does not have an integration".

File: internal/services/rewards.go [hunk 4]
```diff
@@ -229,6 +231,28 @@ func (t *BaselineClient) assignPoints() error {
...
+			if !ok {
+				logger.Warn().Msgf("Synthetic device IntegrationTokenID %d does not have an associated integration.", sd.IntegrationTokenId)
+				continue
+			}
+
+			if integsSignalsThisWeek.Contains(integr.Id) {
+				thisWeek.SyntheticDeviceID = null.IntFrom(int(sd.TokenId))
+				thisWeek.IntegrationPoints += int(integr.Points)
+				thisWeek.IntegrationIds = append(thisWeek.IntegrationIds, integr.Id)
+			}
```
Comment (elffjs) [inline internal/services/rewards.go:250]: This is the heart of it. This is good stuff.

File: internal/services/rewards_test.go [hunk 5]
```diff
@@ -806,6 +806,139 @@ func TestBaselineIssuance(t *testing.T) {
...
+	}()
+
+	scens := []Scenario{
+		{
+			Name: "AutoPiJoinLevel1",
+			Users: []User{
+				{ID: "User1", Address: mkAddr(1)},
+			},
+			Devices: []Device{
+				{ID: mkID(1), TokenID: 1, UserID: "User1", VIN: mkVIN(1), IntsWithData: []string{teslaIntegration}, Beneficiary: mkAddr(2).Bytes(), SDTokenID: 11, SDIntegrationID: 2}},
```
Comment (elffjs) [inline internal/services/rewards_test.go:833]: I don't think beneficiary makes sense here.

File: internal/services/rewards_test.go [hunk 6]
```diff
@@ -806,6 +806,139 @@ func TestBaselineIssuance(t *testing.T) {
...
+
+	cont, conn := utils.GetDbConnection(ctx, t, logger)
+	defer func() {
+		err := cont.Terminate(ctx)
+		assert.NoError(t, err)
+	}()
+
+	scens := []Scenario{
+		{
+			Name: "AutoPiJoinLevel1",
```
Comment (elffjs) [inline internal/services/rewards_test.go:828]: AutoPi?

Comment (elffjs) [reply internal/services/rewards_test.go:828]: It seems like we're mostly checking that things don't explode here, which is good but maybe not at the level of certainty that we'd prefer. Let me think about how to do this—the tests here are gross, I grant that.

File: internal/services/rewards.go (hunk 1 above)
Comment (elffjs) [reply internal/services/rewards.go:256]: And that's a section I think we should delete.

File: internal/services/rewards_test.go [hunk 7]
```diff
@@ -934,7 +934,7 @@ func TestSyntheticDeviceIssuance(t *testing.T) {
 			for _, id := range rw.IntegrationIds {
 				assert.Equal(t, id, id)
 			}
-			assert.Equal(t, rw.IntegrationPoints, 8000)
+			assert.Equal(t, rw.IntegrationPoints, 4000)
```
Comment (elffjs) [inline internal/services/rewards_test.go:937]: I think this is a red flag.

File: internal/storage/db_test.go [hunk 8]
```diff
@@ -114,7 +114,7 @@ func TestTokenAssignmentOneDecrease(t *testing.T) {
 	require.NoError(t, reward1.Insert(context.TODO(), conn.DBS().Writer.DB, boil.Infer()))
 	require.NoError(t, reward2.Insert(context.TODO(), conn.DBS().Writer.DB, boil.Infer()))
 
-	db := DBStorage{DBS: conn}
+	db := DBStorage{DBS: conn, Logger: &logger}
```
Comment (elffjs) [inline internal/storage/db_test.go:117]: My mistake.

File: internal/services/rewards_test.go [hunk 9]
```diff
@@ -918,6 +1053,13 @@ func (d *FakeDevClient) GetUserDevice(_ context.Context, in *pb_devices.GetUserD
 			ud2.AftermarketDeviceTokenId = &ud2.AftermarketDevice.TokenId //nolint:staticcheck
 		}
 
+		if ud.SDTokenID != 0 && ud.SDIntegrationID != 0 {
```
Comment (elffjs) [inline internal/services/rewards_test.go:1056]: Slick.

File: internal/storage/db_test.go (hunk 8 above)
Comment (elffjs) [reply internal/storage/db_test.go:117]: Fixed on main and merged into here.

**Comment by elffjs** [comment]
Looking at the tests, it seems like the software connections aren't validating. I think the same thing you did for your tests, where cars have optional synthetic device ids and integration ids, we'll have to repeat here.

File: internal/services/rewards.go [hunk 10]
```diff
@@ -229,15 +229,22 @@ func (t *BaselineClient) assignPoints() error {
...
-				break
+		if sd := ud.SyntheticDevice; sd != nil {
+			if sd.IntegrationTokenId == 0 {
+				logger.Warn().Msg("synthetic device does not have an integration")
+				continue
+			}
+
+			integr, ok := swIntegrsByTokenID[sd.IntegrationTokenId]
+			if !ok {
+				logger.Warn().Msgf("Sunknown integration with token id %d", sd.IntegrationTokenId)
```
Comment (Allyson-English) [inline internal/services/rewards.go:240]:
```suggestion
				logger.Warn().Msgf("unknown synthetic integration with token id %d", sd.IntegrationTokenId)
```

File: internal/services/rewards.go [hunk 11]
```diff
@@ -139,12 +139,12 @@ func (t *BaselineClient) assignPoints() error {
 	}
 
 	amMfrTokenToIntegration := make(map[uint64]*pb_defs.Integration)
-	swIntegrsByID := make(map[string]*pb_defs.Integration)
+	swIntegrsByTokenID := make(map[uint64]*pb_defs.Integration)
 
 	for _, integr := range allIntegrations.Integrations {
 		if integr.ManufacturerTokenId == 0 {
 			// Must be a software integration. Sort after this loop.
```
Comment (Allyson-English) [inline internal/services/rewards.go:146]:
```suggestion
			// Must be a software integration.
```

File: internal/services/rewards_test.go [hunk 12]
```diff
@@ -806,6 +806,139 @@ func TestBaselineIssuance(t *testing.T) {
...
+
+	logger := zerolog.Nop()
+
+	cont, conn := utils.GetDbConnection(ctx, t, logger)
+	defer func() {
+		err := cont.Terminate(ctx)
+		assert.NoError(t, err)
+	}()
+
+	scens := []Scenario{
```
Comment (Allyson-English) [inline internal/services/rewards_test.go:826]: suggestion:
- add test that includes different combinations:
          - two synthetic devices --> only connection should earn
          - synthetic and aftermarket --> if valid pairing (ie not tesla and autopi), earn expected total

**Comment by elffjs** [comment]
Seems like the test containers are just being weird.

File: internal/services/rewards_test.go [hunk 13]
```diff
@@ -253,32 +253,13 @@ func TestStreak(t *testing.T) {
 			},
 			NewVIN: []VIN{},
 		},
-		{
```
Comment (elffjs) [inline internal/services/rewards_test.go:256]: Must we delete this? What goes wrong with it?

File: internal/services/rewards_test.go [hunk 14]
```diff
@@ -847,10 +982,10 @@ type FakeDefClient struct {
...
 func (d *FakeDefClient) GetIntegrations(_ context.Context, _ *emptypb.Empty, _ ...grpc.CallOption) (*pb_defs.GetIntegrationResponse, error) {
 	return &pb_defs.GetIntegrationResponse{Integrations: []*pb_defs.Integration{
-		{Id: autoPiIntegration, Vendor: "AutoPi", ManufacturerTokenId: 137, Points: 6000},
-		{Id: teslaIntegration, Vendor: "Tesla", Points: 4000, ManufacturerTokenId: 0},
-		{Id: smartcarIntegration, Vendor: "SmartCar", Points: 1000, ManufacturerTokenId: 0},
-		{Id: macaronIntegration, Vendor: "Macaron", ManufacturerTokenId: 142, Points: 2000},
+		{Id: autoPiIntegration, Vendor: "AutoPi", ManufacturerTokenId: 137, Points: 6000, TokenId: 1},
+		{Id: teslaIntegration, Vendor: "Tesla", Points: 4000, ManufacturerTokenId: 0, TokenId: 2},
+		{Id: smartcarIntegration, Vendor: "SmartCar", Points: 1000, ManufacturerTokenId: 0, TokenId: 3},
+		{Id: macaronIntegration, Vendor: "Macaron", ManufacturerTokenId: 142, Points: 2000, TokenId: 4},
```
[the remark this answers is not in the document]
Comment (elffjs) [reply internal/services/rewards_test.go:969]: Er, sorry, I meant AutoPi and Macaron. That's my mistake. Typing too fast.

[the remark this answers is not in the document]
Comment (elffjs) [reply internal/services/rewards_test.go]: I removed the token id from the Macaron.

File: internal/services/rewards_test.go (hunk 12 above)
Comment (elffjs) [reply internal/services/rewards_test.go:807]: I think we've covered your second scenario in the above suite. The first one would be good to see—can we add such a test? Something in which we've seen signals from both Smartcar and Tesla, but only the Tesla synthetic device is paired.

To be honest, it's not clear to me that this should be in a separate suite. Is the shape any different from the earlier tests?

**Review by elffjs — COMMENTED** [review]
Remaining bits from me:

1. Do we need this separate suite? Can we dump the Kafka stuff that we're testing elsewhere?
2. Do we need to delete this test we've deleted?
3. Can we add a scenario in which multiple software integrations have transmitted but only one is currently paired as a synthetic device? That was a gap in the old suite.

Overall it looks very good, the logic is crisp.

**Review by elffjs — APPROVED** [review]

**Pull request merged**